src/lib/storage/table.cpp

```diff
@@ -71,6 +72,16 @@ Chunk& Table::get_chunk(ChunkID chunk_id) { return _chunks.at(chunk_id); }
 
 const Chunk& Table::get_chunk(ChunkID chunk_id) const { return _chunks.at(chunk_id); }
 
-void Table::compress_chunk(ChunkID chunk_id) { throw std::runtime_error("Implement Table::compress_chunk"); }
+void Table::compress_chunk(ChunkID chunk_id) {
+  auto new_chunk = Chunk();
+
+  for (auto segment_id = ColumnID{0}; segment_id < column_count(); segment_id++) {
+    auto dict_segment = opossum::make_shared_by_data_type<opossum::BaseSegment, opossum::DictionarySegment>(
+        column_type(segment_id), _chunks[chunk_id].get_segment(segment_id));
+    new_chunk.add_segment(dict_segment);
+  }
+
+  _chunks[chunk_id] = std::move(new_chunk);
```

Comment:
Implement multithreading. I would do it like this - compiles and should work, but I did not do extensive testing.
```
 std::vector<std::thread> threads;
  threads.reserve(chunk.column_count());

  for (ColumnID current_column_id{0}; current_column_id < chunk.column_count(); current_column_id++) {
    threads.emplace_back(
      [&compressed_chunk, &chunk, current_column_id, column_type = column_type(current_column_id)] {
        auto dictionary_segment =
          make_shared_by_data_type<BaseSegment, DictionarySegment>(column_type,
                                                                   chunk.get_segment(current_column_id));

        compressed_chunk.add_segment(dictionary_segment);
      });
  }

  for (auto & thread : threads)
    thread.join();
```

Reply:
```
  Chunk compressed_chunk;
  const Chunk& current_chunk = get_chunk(chunk_id);

  std::vector<std::thread> threads;
  std::vector<std::future<std::shared_ptr<BaseSegment>>> futures;
  threads.reserve(_column_types.size());
  futures.reserve(_column_types.size());

  // compressing one chunks means turning all the ValueSegments into DictionarySegments
  for (std::size_t i = 0; i < _column_types.size(); i++) {
    std::promise<std::shared_ptr<BaseSegment>> p;
    futures.push_back(p.get_future());
    threads.push_back(
        std::thread(compress_segment, _column_types[i], current_chunk.get_segment(ColumnID(i)), std::move(p)));
  }

  for (std::size_t i = 0; i < _column_types.size(); i++) {
    threads[i].join();
    compressed_chunk.add_segment(futures[i].get());
  }

  // then switch the current chunk with the newly compressed one
  std::unique_lock<std::mutex> lock(_mutex);
  auto it = _chunks.erase(_chunks.begin() + chunk_id);
  it = _chunks.insert(it, std::move(compressed_chunk));
  lock.unlock();
```